smt.go

```diff
@@ -357,31 +360,68 @@ func (smt *SparseMerkleTree) sideNodesForRoot(path []byte, root []byte) ([][]byt
 
 		if bytes.Equal(nodeHash, smt.th.placeholder()) {
 			// If the node is a placeholder, we've reached the end.
-			return reverseSideNodes(sideNodes), nodeHash, nil, nil
+			if getSiblingData {
+				siblingData, err = smt.ms.Get(sideNodes[i])
+				if err != nil {
+					return nil, nil, nil, nil, err
+				}
+			}
+			return reverseSideNodes(sideNodes), nodeHash, nil, siblingData, nil
 		}
 
 		currentData, err = smt.ms.Get(nodeHash)
 		if err != nil {
-			return nil, nil, nil, err
+			return nil, nil, nil, nil, err
 		} else if smt.th.isLeaf(currentData) {
 			// If the node is a leaf, we've reached the end.
+			if getSiblingData {
+				siblingData, err = smt.ms.Get(sideNodes[i])
+				if err != nil {
+					return nil, nil, nil, nil, err
+				}
+			}
 			break
 		}
 	}
 
-	return reverseSideNodes(sideNodes), nodeHash, currentData, nil
+	return reverseSideNodes(sideNodes), nodeHash, currentData, siblingData, nil
```

Comment:
In line 383, `break` is used to exit the loop and go to `return reverseSideNodes(sideNodes), nodeHash, currentData, siblingData, nil` line.
In line 369, almost the same `return` statement is used (with `currentData = nil`). 
It's 7 lines of duplicated code, and it's looks ugly because it's deeply nested. Introducing a function to "wrap" this code is a bit of overkill IMHO.
I can't make a suggestion for such a big change, so I'll posted small PR targeting your branch, that simplifies the entire loop: https://github.com/lazyledger/smt/pull/29.

Reply:
Merged, thanks!